File: golang/54_spiral_matrix.go
func spiralOrder(matrix [][]int) []int {
	var res []int
	m := len(matrix)
	if m == 0 {
		return res
	}
	n := len(matrix[0])
	k, l := 0, 0
	for k < m && l < n {
		for i := l; i < n; i++ {
			res = append(res, matrix[k][i])
		}
		k++
		for i := k; i < m; i++ {
			res = append(res, matrix[i][n-1])
		}
		n--
		if k < m {
			for i := n - 1; i > l-1; i-- {
				res = append(res, matrix[m-1][i])
			}
			m--
		}
		if l < n {
			for i := m - 1; i > k-1; i-- {
				res = append(res, matrix[i][l])
			}
			l++
		}
	}
	return res
}

File: python/54_spiral_matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        res = []
        m = len(matrix)
        if m == 0:
            return []
        n = len(matrix[0])
        k = 0; l = 0
        while (k < m and l < n) : 
            for i in range(l, n) : 
                res.insert(len(res), matrix[k][i])
    
            k += 1
            for i in range(k, m) : 
                res.insert(len(res), matrix[i][n - 1]) 
              
            n -= 1
            if ( k < m) : 
                for i in range(n - 1, (l - 1), -1) : 
                    res.insert(len(res), matrix[m - 1][i]) 
                m -= 1
            if (l < n) : 
                for i in range(m - 1, k - 1, -1) : 
                    res.insert(len(res), matrix[i][l]) 
                l += 1
        return res